### PR TITLE
Gtfs benchmarks

### PR DESCRIPTION
**Pull request by austinjpaul**
Add code to process GTFS data to create benchmark headways and traveltimes for bus routes.

Code is live on beta, and the benchmark data exists for all bus routes for June and July 2022.
Check it out there and make sure things look reasonable!
I still need to figure out how to disable it for 114-116-117 somehow, since those headways aren't for the combined route.

### CODE REVIEW

**Comment by austinjpaul** [comment]
#205 was no good, so reopened with a new approach.
This will address #116

File: server/bus/gtfs_archive.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import shutil
+import urllib.request
+
+MAIN_DIR = pathlib.Path("./data/gtfs_archives/")
+MAIN_DIR.mkdir(parents=True, exist_ok=True)
+
+ARCHIVES = pd.read_csv("https://cdn.mbta.com/archive/archived_feeds.txt")
+
+
+def to_dateint(date):
```
Comment (mathcolo) [inline server/bus/gtfs_archive.py:13]: Ehhhhhh :( Dates aren't integers, so we probably shouldn't pretend like they are.

File: server/bus/gtfs_archive.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    Determine which GTFS archive corresponds to the date.
+    Returns that archive folder, downloading if it doesn't yet exist.
+    """
+    matches = ARCHIVES[(ARCHIVES.feed_start_date <= dateint) & (ARCHIVES.feed_end_date >= dateint)]
+    archive_url = matches.iloc[0].archive_url
+
+    archive_name = pathlib.Path(archive_url).stem
+
+    if (MAIN_DIR / archive_name).exists():
```
Comment (mathcolo) [inline server/bus/gtfs_archive.py:28]: Is `/` really overloaded to work as an append for Paths? WTF

File: server/bus/gtfs_archive.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    cal = pd.read_csv(archive_dir / "calendar.txt")
+    current_services = cal[(cal.start_date <= dateint) & (cal.end_date >= dateint)]
+    services = current_services[current_services[day_of_week] == 1].service_id.tolist()
+
+    exceptions = pd.read_csv(archive_dir / "calendar_dates.txt")
+    exceptions = exceptions[exceptions.date == dateint]
+    additions = exceptions[exceptions.exception_type == 1].service_id.tolist()
+    subtractions = exceptions[exceptions.exception_type == 2].service_id.tolist()
+
+    services = (set(services) - set(subtractions)) | set(additions)
```
Comment (mathcolo) [inline server/bus/gtfs_archive.py:59]: Wow this is elegant. nice

File: server/bus/gtfs_archive.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        gtfs_stops['scheduled_headway'] = headways.dt.seconds
+
+        # calculate gtfs traveltimes
+        trip_start_times = gtfs_stops.groupby('trip_id').arrival_time.transform('min')
+        gtfs_stops['scheduled_tt'] = (gtfs_stops.arrival_time - trip_start_times).dt.seconds
+
+        # assign each actual timepoint a scheduled headway
+        # merge_asof 'backward' matches the previous scheduled value of 'arrival_time'
+        days_events['arrival_time'] = days_events.event_time - service_date
+        final = pd.merge_asof(days_events.sort_values(by='arrival_time'),
```
Comment (mathcolo) [inline server/bus/gtfs_archive.py:131]: Oh... that is amazing.

File: server/chalicelib/s3_historical.py [hunk 5]
```diff
@@ -83,7 +83,7 @@ def headways(stop_ids, sdate, edate):
             "direction": this["direction_id"],
             "current_dep_dt": this["event_time"],
             "headway_time_sec": headway_time_sec,
-            "benchmark_headway_time_sec": None
+            "benchmark_headway_time_sec": this.get("scheduled_headway")
```
Comment (mathcolo) [inline server/chalicelib/s3_historical.py:86]: Are you using `get()` for a reason here?

File: server/chalicelib/s3_historical.py [hunk 6]
```diff
@@ -116,13 +116,21 @@ def travel_times(stops_a, stops_b, sdate, edate):
         if travel_time_sec < 0:
             continue
 
+        # benchmark calculation:
+        sched_arr = arrival.get("scheduled_tt")
+        sched_dep = departure.get("scheduled_tt")
+        if sched_arr is not None and sched_dep is not None:
```
Comment (mathcolo) [inline server/chalicelib/s3_historical.py:122]: Can this be `benchmark = float(sched_arr) - float(sched_dep) or None`?

File: server/bus/gtfs_archive.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    from gtfs for the routes we care about, and then match our actual
+    events to the scheduled values. This matching is done based on
+    time-of-day, so is not an excact match. Luckily, pandas helps us out
+    with merge_asof.
+    https://pandas.pydata.org/docs/reference/api/pandas.merge_asof.html
+    """
+    # TODO: I think we need to worry about 114/116/117 headways?
+
+    # defining these columns in particular becasue we use them everywhere
+    RDS = ['route_id', 'direction_id', 'stop_id']
```
Comment (mathcolo) [inline server/bus/gtfs_archive.py:104]: Can we change this to `KEY_COLUMNS` or something?

**Review by mathcolo — APPROVED** [review]
Algorithmically looks good, some nitpicks. Nice!

File: server/bus/gtfs_archive.py (hunk 1 above)
Comment (austinjpaul) [reply server/bus/gtfs_archive.py:13]: hmm... I was just trying to match the gtfs csv (https://cdn.mbta.com/archive/archived_feeds.txt).
I suppose I can instead specify `dtypes` when loading all the gtfs files, if that would be preferable?

File: server/bus/gtfs_archive.py (hunk 2 above)
Comment (austinjpaul) [reply server/bus/gtfs_archive.py:28]: One of the many delights of working with pathlib! https://docs.python.org/3/library/pathlib.html#operators

File: server/chalicelib/s3_historical.py (hunk 5 above)
Comment (austinjpaul) [reply server/chalicelib/s3_historical.py:86]: Not all files will have "benchmark_headway_time_sec" as a defined column (rail archives, bus archives that haven't been updated yet, bus routes that may not be eligible for this treatment)

**Comment by austinjpaul** [comment]
#225 has been addressed by https://github.com/transitmatters/t-performance-dash/pull/208/commits/1b6c5c3a9bdef2f16ef6fcabf2388a76121f047a

File: server/chalicelib/s3_historical.py (hunk 6 above)
Comment (austinjpaul) [reply server/chalicelib/s3_historical.py]: fixed

File: server/bus/gtfs_archive.py (hunk 1 above)
Comment (austinjpaul) [reply server/bus/gtfs_archive.py:13]: I've decided that it will be way too obnoxious to try and impose rigorous date logic on gtfs.

**Comment by mathcolo** [comment]
could you bring this up to date with `main`?

**Comment by mathcolo** [comment]
oh the GitHub UI confused me, sorry! 6668602 was today, not "3 months ago" lol

**Pull request merged**